datafusion/datasource-parquet/src/metadata.rs

```diff
@@ -0,0 +1,564 @@
+};
+use datafusion_execution::cache::cache_manager::{FileMetadata, FileMetadataCache};
+use datafusion_functions_aggregate::min_max::{MaxAccumulator, MinAccumulator};
+use datafusion_physical_plan::Accumulator;
+use log::debug;
+use object_store::path::Path;
+use object_store::{ObjectMeta, ObjectStore};
+use parquet::arrow::arrow_reader::statistics::StatisticsConverter;
+use parquet::arrow::parquet_to_arrow_schema;
+use parquet::file::metadata::{ParquetMetaData, ParquetMetaDataReader, RowGroupMetaData};
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// Handles fetching Parquet file schema, metadata and statistics
+/// from object store.
+///
+/// This component is exposed for low level integrations through
+/// [`ParquetFileReaderFactory`].
+///
+/// [`ParquetFileReaderFactory`]: crate::ParquetFileReaderFactory
+#[derive(Debug)]
+pub struct DFParquetMetadata<'a> {
+    store: &'a dyn ObjectStore,
+    object_meta: &'a ObjectMeta,
+    metadata_size_hint: Option<usize>,
+    decryption_properties: Option<&'a FileDecryptionProperties>,
+    file_metadata_cache: Option<Arc<dyn FileMetadataCache>>,
+    /// timeunit to coerce INT96 timestamps to
+    pub coerce_int96: Option<TimeUnit>,
+}
+
+impl<'a> DFParquetMetadata<'a> {
+    pub fn new(store: &'a dyn ObjectStore, object_meta: &'a ObjectMeta) -> Self {
+        Self {
+            store,
+            object_meta,
+            metadata_size_hint: None,
+            decryption_properties: None,
+            file_metadata_cache: None,
+            coerce_int96: None,
+        }
+    }
+
+    /// set metadata size hint
+    pub fn with_metadata_size_hint(mut self, metadata_size_hint: Option<usize>) -> Self {
+        self.metadata_size_hint = metadata_size_hint;
+        self
+    }
+
+    /// set decryption properties
+    pub fn with_decryption_properties(
+        mut self,
+        decryption_properties: Option<&'a FileDecryptionProperties>,
+    ) -> Self {
+        self.decryption_properties = decryption_properties;
+        self
+    }
+
+    /// set file metadata cache
+    pub fn with_file_metadata_cache(
+        mut self,
+        file_metadata_cache: Option<Arc<dyn FileMetadataCache>>,
+    ) -> Self {
+        self.file_metadata_cache = file_metadata_cache;
+        self
+    }
+
+    /// Set timeunit to coerce INT96 timestamps to
+    pub fn with_coerce_int96(mut self, time_unit: Option<TimeUnit>) -> Self {
+        self.coerce_int96 = time_unit;
+        self
+    }
+
+    /// fetch parquet metadata
+    pub async fn fetch_metadata(&self) -> Result<Arc<ParquetMetaData>> {
+        let Self {
+            store,
+            object_meta,
+            metadata_size_hint,
+            decryption_properties,
+            file_metadata_cache,
+            coerce_int96: _,
+        } = self;
+
+        let fetch = ObjectStoreFetch::new(*store, object_meta);
+
+        // implementation to fetch parquet metadata
```

Comment:
I'm not reviewing the implementation, I assume it was largely copied over

Reply:
Yes, it was entirely copied over